registry/api/errcode/errors.go

```diff
@@ -0,0 +1,206 @@
+	// this error condition.
+	HTTPStatusCode int
+}
+
+var (
+	errorCodeToDescriptors = map[ErrorCode]ErrorDescriptor{}
+	idToDescriptors        = map[string]ErrorDescriptor{}
+)
+
+const (
+	// ErrorCodeUnknown is a catch-all for errors not defined below.
+	ErrorCodeUnknown ErrorCode = 10000 + iota
+)
+
+var errorDescriptors = []ErrorDescriptor{
+	{
+		Code:    ErrorCodeUnknown,
+		Value:   "UNKNOWN",
+		Message: "unknown error",
+		Description: `Generic error returned when the error does not have an
+									        API classification.`,
+		HTTPStatusCode: http.StatusInternalServerError,
+	},
+}
+
+// LoadErrors will register a new set of Errors into the system
+func LoadErrors(errs *[]ErrorDescriptor) {
+	for _, descriptor := range *errs {
+		if _, ok := idToDescriptors[descriptor.Value]; ok {
+			panic(fmt.Sprintf("ErrorValue %s is already registered", descriptor.Value))
+		}
+		if _, ok := errorCodeToDescriptors[descriptor.Code]; ok {
+			panic(fmt.Sprintf("ErrorCode %d is already registered", descriptor.Code))
+		}
+
+		errorCodeToDescriptors[descriptor.Code] = descriptor
+		idToDescriptors[descriptor.Value] = descriptor
+	}
+}
+
+// ParseErrorCode attempts to parse the error code string, returning
+// ErrorCodeUnknown if the error is not known.
+func ParseErrorCode(s string) ErrorCode {
+	desc, ok := idToDescriptors[s]
+
+	if !ok {
+		return ErrorCodeUnknown
+	}
+
+	return desc.Code
+}
+
+// Descriptor returns the descriptor for the error code.
+func (ec ErrorCode) Descriptor() ErrorDescriptor {
+	d, ok := errorCodeToDescriptors[ec]
+
+	if !ok {
+		return ErrorCodeUnknown.Descriptor()
+	}
+
+	return d
+}
+
+// String returns the canonical identifier for this error code.
+func (ec ErrorCode) String() string {
+	return ec.Descriptor().Value
+}
+
+// Message returned the human-readable error message for this error code.
+func (ec ErrorCode) Message() string {
+	return ec.Descriptor().Message
+}
+
+// MarshalText encodes the receiver into UTF-8-encoded text and returns the
+// result.
+func (ec ErrorCode) MarshalText() (text []byte, err error) {
+	return []byte(ec.String()), nil
+}
+
+// UnmarshalText decodes the form generated by MarshalText.
+func (ec *ErrorCode) UnmarshalText(text []byte) error {
+	desc, ok := idToDescriptors[string(text)]
+
+	if !ok {
+		desc = ErrorCodeUnknown.Descriptor()
+	}
+
+	*ec = desc.Code
+
+	return nil
+}
+
+// Error provides a wrapper around ErrorCode with extra Details provided.
+type Error struct {
+	Code    ErrorCode   `json:"code"`
+	Message string      `json:"message,omitempty"`
+	Detail  interface{} `json:"detail,omitempty"`
+}
+
+// Error returns a human readable representation of the error.
+func (e Error) Error() string {
+	return fmt.Sprintf("%s: %s",
+		strings.ToLower(strings.Replace(e.Code.String(), "_", " ", -1)),
+		e.Message)
+}
+
+// Errors provides the envelope for multiple errors and a few sugar methods
+// for use within the application.
+type Errors struct {
```

Comment:
Let's change this typedef to be a slice:

```
type Errors []error
```

This makes this work generally with `append`, `len` and other slice operations. It should just hold generic errors and type assert them to the correct types for serialization. `UnmarshalJSON/MarshalJSON` will a struct definition to handle the serialization and deserialization, accordingly. `Push` and `PushErr` can be removed. `Push` can be replaced with functionality on `ErrorCode` that let's one add detail and format the error message (maybe `(ErrorCode).Include(msg, detail ...interface{}) Error`).